docs/lab/lab04/lab04.md

```diff
@@ -0,0 +1,384 @@
+:::
+
+![ADCSampling](images/sampling_values.svg)
+
+### Nyquist-Shannon Sampling Theorem
+
+For an analog signal to be represented without loss of information, the conversion needs to satisfy the following formula:
+
+$$
+sampling_f >= 2 \times max_{f}
+$$
+
+The analog signal needs to be sampled at a frequency greater than twice the *maximum frequency* of the signal.
+In other words, we must sample at least twice per cycle.
+
+![NyquistTheorem](images/nyquist_theorem.png)
+
+### Example of analog sensor
+
+A **photoresistor** (or photocell) is a sensor that measures the intensity of light around it. Its internal resistance varies depending on the light hitting its surface, therefore the more light there is, the lower the resistance will be. 
+
+![Photoresistor](images/photoresistor.png)
+
+### ADC in Embassy-rs
+
+On the RP2040, ADC uses an interrupt called `ADC_IRQ_FIFO` to signal whenever a new sample has been processed. This new sample will be stored inside a FIFO. In the Embassy library, this interrupt is already implemented, so all we need to do is bind it and use it in our ADC variable.
+
+```rust
+bind_interrupts!(struct Irqs {
+    ADC_IRQ_FIFO => InterruptHandler;
+});
+
+// ---- fn main() ----
+
+// Initialize peripherals
+let peripherals = embassy_rp::init(Default::default());
+
+// Create ADC driver
+let mut adc = Adc::new(peripherals.ADC, Irqs, Config::default());
+```
+
+:::warning
+If we are using PWM and ADC in the same code, we will have two different `Config` imports with the same name. In order to avoid compilation errors, we need to separate the PWM config import from the ADC one. To do this, we can import the two `Config`s with different names.
+
+```rust
+use embassy_rp::pwm::Config as ConfigPwm; // PWM config
+use embassy_rp::adc::Config as ConfigAdc; // ADC config
+
+// ---- fn main() ----
+let mut adc = Adc::new(peripherals.ADC, Irqs, ConfigAdc::default());
+```
+:::
+
+Now, we need to initialize the ADC pin we will be using. The Raspberry Pi Pico has 3 pins that support ADC: ADC0, ADC1, and ADC2.
+
+```rust
+// Initialize ADC pin
+let mut adc_pin = Channel::new_pin(peripherals.PIN_X, Pull::None); // where X should be replaced with a pin number
+```
+
+Once we have the ADC and pin set up, we can start reading values from the pin. 
+
+```rust
+let level = adc.read(&mut adc_pin).await.unwrap(); // read a value from the pin
+info!("Light sensor reading: {}", level); // print the value over serial
+Timer::after_secs(1).await; // wait a bit before reading and printing another value
+```
+
+## Using channels in Embassy
+
+Up to this point, to be able to share peripherals across multiple tasks, we have been using global `Mutex`s. But there are other, more convenient ways to send data to and from tasks. Instead of having to make global, static variables that are shared by tasks, we could choose to only send the information that we need from one task to another. To achieve this, we can use *channels*.
```

Comment:
In embassy we used parameters to a task.

Reply:
Right, I encountered lifetime issues when trying to pass PWM to multiple tasks. That's why I figured that this case calls for Mutex or for channels. Removed this whole bit anyway, and I'll come back to it for next week's lab.